### ISSUE TITLE
Code Editor line spacing too large even at zero

### ISSUE THREAD

**Issue by amonroejj** [godotengine/godot#76385]
### Godot version

4.0.2.stable

### System information

Windows 10, NVIDIA GeForce GTX 1060 516.94

### Issue description

Godot 3.x: line spacing 2 (can get even tighter at 1 or 0!)
![image](https://user-images.githubusercontent.com/2057161/233866392-d30d1780-e5bd-41d5-82b5-7414cc060ac8.png)

Godot 4.x: line spacing 0 (not dense enough, and cannot get any tighter!)
![image](https://user-images.githubusercontent.com/2057161/233866410-b7547bb8-dcb0-40d7-b584-2be2c70f7fc9.png)

### Steps to reproduce

![image](https://user-images.githubusercontent.com/2057161/233866539-ca2ddc95-856e-4ef9-ae20-bb4b965cd7a6.png)

### Minimal reproduction project

N/A

**Comment by Chaosus** [comment]
It may be depended on fonts used by the editor. In 3.x default font was `Hack`, in 4.x (as far as I remember) it used `Jetbrains Mono`.

**Comment by Rindbee** [comment]
It may also be due to #74528 that the value change did not take effect.

**Comment by Calinou** [comment]
> It may be depended on fonts used by the editor. In 3.x default font was `Hack`, in 4.x (as far as I remember) it used `Jetbrains Mono`.

@amonroejj is using a custom font in the screenshots, although they didn't specify which font it is (nor the font size used).

**Comment by amonroejj** [comment]
Font is ProFontWindows.ttf from https://tobiasjung.name/profont/ . Code font size is 12.

**Comment by Calinou** [comment]
@amonroejj Did you try restarting the editor after changing the setting?

I guess the way line spacing is calculated has changed in 4.0 due to the TextServer refactoring. We could allow setting the line spacing setting as low as `-4` to resolve this.

**Comment by amonroejj** [comment]
@Calinou Yes, the increased line spacing persists over relaunch & reboot.

**Comment by Calinou** [comment]
While the editor is closed, try opening [`editor_settings-4.tres`](https://docs.godotengine.org/en/stable/tutorials/io/data_paths.html#editor-data-paths) and change `text_editor/appearance/whitespace/line_spacing`'s value to something like `-4` to see if it works.

**Comment by amonroejj** [comment]
One offshoot of this is that the hotspot of the I-beam mouse cursor seems not to be centered vertically.

![image](https://user-images.githubusercontent.com/2057161/235375377-0a7377b9-2d07-43bd-9f22-6f8f1a97f2de.png)

With the mouse pointer in this position, the expectation would be to put the cursor in the fourth line of visible text, but as seen in the screenshot, the actual result is that the cursor goes to the third line of visible text.

**Comment by Rindbee** [comment]
> One offshoot of this is that the hotspot of the I-beam mouse cursor seems not to be centered vertically, so mouse click doesn't put the cursor where expected.
> 
> ![image](https://user-images.githubusercontent.com/2057161/235375377-0a7377b9-2d07-43bd-9f22-6f8f1a97f2de.png)
> 
> With the mouse pointer in this position, the expectation would be to put the cursor in the fourth line of visible text, but as seen in the screenshot, the actual result is that the cursor goes to the third line of visible text.

Seems to be a bug in `TextEdit`. The mouse click to select the line is not only related to `get_line_height()`, but also related to the height of caret (or `font_size`).

![0](https://user-images.githubusercontent.com/30386067/236711086-21ebdad8-4157-4809-8462-0cf3c693981f.gif)

**Comment by amonroejj** [comment]
Something I just noticed regarding line spacing is that that the Output pane (green highlight) perfectly copes with tight line spacing while the code editor (yellow highlight) does not:

![image](https://github.com/godotengine/godot/assets/2057161/d09b7c7f-dc93-49ab-b428-4e6f18da33b6)

I'm not familiar enough with the internals of the editor to know whether different classes are used between these two controls in the GUI.

**Comment by enetheru** [comment]
I was reading the source this afternoon, 
`TextEdit._notification() ... NOTIFICATION_DRAW :`  uses `row_height` retrieved from `TextEdit.get_line_height()`
This in turn calls `text.get_line_height()` which loops over the lines of text, and takes the maximum of their heights.

https://github.com/godotengine/godot/blob/4e6ac9bccc7868432b9d8a009d34860a2e6f3393/scene/gui/text_edit.cpp#L137

These `Text::Line.height` values are set when `Text::invalidate_cache` is called.

https://github.com/godotengine/godot/blob/4e6ac9bccc7868432b9d8a009d34860a2e6f3393/scene/gui/text_edit.cpp#L362

Which unfortunately also loops over lines and takes the maximum of `Text.font_height` or the line height reported from the `Text::Line.TextParagraph.get_line_size` 

https://github.com/godotengine/godot/blob/4e6ac9bccc7868432b9d8a009d34860a2e6f3393/scene/resources/text_paragraph.cpp#L771

Anyway, this rabbit hole of chasing leads to text server `return TS->shaped_text_get_size(lines_rid[p_line]);`

Which for a standard editor leads to the text_server_adv, 

https://github.com/godotengine/godot/blob/4e6ac9bccc7868432b9d8a009d34860a2e6f3393/modules/text_server_adv/text_server_adv.cpp#L7263

Which is why if you set the spacing on the font, you can sort of mess with it to remove the additional height added, but that it only works for a specific zoom level.

**Comment by enetheru** [comment]
here's a little snippet to sort of help mitigate the struggle:
```gdscript
@tool
extends EditorScript

func _run() -> void:
	var editor_theme : Theme = EditorInterface.get_editor_theme()
	var code_edit_font : FontVariation = editor_theme.get_font("font", "CodeEdit")
	code_edit_font.spacing_bottom = -3 # Original is -1
	code_edit_font.spacing_top = -4 # Original is -1

```
Adjust the negative spacing values to suit your font and zoom level, this as fragile as it comes, reloading the editor, or changing the font or zoom level immediately ruins it.

**Comment by dfcut8** [comment]
Same issue. I like condensed fonts, Tried to the following fonts and line spacing looks bad:
- Iosevka
- MPlus
- Agave

In Godot Engine v4.4.1.stable.official:
<img width="802" height="136" alt="Image" src="https://github.com/user-attachments/assets/06293e42-c9f1-4008-8314-9dfb0e26df70" />

In vscode:
<img width="678" height="117" alt="Image" src="https://github.com/user-attachments/assets/382dba4d-9b7e-4d14-9198-2be2f4aa4ea7" />
